cvmfs/export_plugin/fs_traversal.cc

```diff
@@ -0,0 +1,793 @@
+      return getNext(fs, dir, dir_list, entry, iter,
+                     st, get_hash, is_src, pstats);
+    }
+    if (entry && !updateStat(fs, *entry, st, get_hash)) {
+      return getNext(fs, dir, dir_list, entry, iter,
+                     st, get_hash, is_src, pstats);
+    }
+  } else {
+    return false;
+  }
+  if (is_src) {
+    pstats->Lookup(SHRINKWRAP_STAT_SRC_ENTRIES)->Inc();
+  } else {
+    pstats->Lookup(SHRINKWRAP_STAT_DEST_ENTRIES)->Inc();
+  }
+  return true;
+}
+
+void list_src_dir(
+  struct fs_traversal *src,
+  const char *dir,
+  char ***buf,
+  size_t *len
+) {
+  int retval = spec_tree_->ListDir(dir, buf, len);
+
+  if (retval == SPEC_READ_FS) {
+    src->list_dir(src->context_, dir, buf, len);
+    qsort(*buf, *len, sizeof(char *),  strcmp_list);
+  }
+}
+
+/**
+ * This method checks whether a file has been manually modified by the source
+ * system and therefore needs to be rewritten (based on is_hash_consistent)
+ * 
+ * This method will return true exactly once for each inode and can therefore
+ * be used as an atomic locking procedure for file writing decisions.
+ */
+bool should_write_anyway(struct fs_traversal *dest,
+  struct cvmfs_attr *src_st,
+  struct cvmfs_attr *dest_st) {
+  if (!dest->is_hash_consistent(dest->context_, dest_st)
+    && fsck_lock->AddLock(dest_st->st_ino)) {
+    return true;
+  }
+  return false;
+}
+
+bool handle_file(
+  struct fs_traversal *src,
+  struct cvmfs_attr *src_st,
+  struct fs_traversal *dest,
+  struct cvmfs_attr *dest_st,
+  const char *entry,
+  perf::Statistics *pstats
+) {
+  bool result = true;
+  // They don't point to the same data, link new data
+  char *dest_data = dest->get_identifier(dest->context_, src_st);
+
+  // Touch is atomic, if it fails something else will write file
+  if (!dest->touch(dest->context_, src_st)
+    || (dest_st->cvm_checksum != NULL
+      && should_write_anyway(dest, src_st, dest_st)) ) {
+    char *src_ident = src->get_identifier(src->context_, src_st);
+    if (num_parallel_) {
+      FileCopy next_copy(src_ident, strdup(dest_data));
+
+      WritePipe(pipe_chunks[1], &next_copy, sizeof(next_copy));
+      atomic_inc64(&copy_queue);
+    } else {
+      if (!copyFile(src, src_ident, dest, dest_data, pstats)) {
+        LogCvmfs(kLogCvmfs, kLogStderr,
+          "Failed to copy %s->%s : %d : %s",
+          entry, dest_data, errno, strerror(errno));
+        errno = 0;
+        result = false;
+      }
+      free(src_ident);
+      pstats->Lookup(SHRINKWRAP_STAT_FILE_COUNT)->Inc();
+    }
+  } else {
+    pstats->Lookup(SHRINKWRAP_STAT_DEDUPED_FILES)->Inc();
+    pstats->Lookup(SHRINKWRAP_STAT_DEDUPED_BYTES)->Xadd(src_st->st_size);
+  }
+
+  // Should probably happen in the copy function as it is parallel
+  // Also this needs to be separate from copyFile, the target file
+  // could already exist and the link needs to be created anyway.
+  if (result && dest->do_link(dest->context_, entry, dest_data)) {
```

Comment:
Clean up this comment and clarify that the file will exist as a result of the touch (either it already existed or was created). The link creation is separate from copy so that updating where the link points is not tied to writing the file. I think what I was trying to communicate is that this doesn't happen in parallel(on a thread). Should it?

Reply:
Shortening comment.